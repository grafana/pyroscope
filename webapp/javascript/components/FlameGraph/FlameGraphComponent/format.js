/* eslint-disable no-plusplus */
/* eslint-disable prefer-destructuring */
/* eslint-disable max-classes-per-file */
export function numberWithCommas(x) {
  return x.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ",");
}

const suffixes = ["K", "M", "G", "T"];

export function shortNumber(x) {
  let suffix = "";

  for (let i = 0; x > 1000 && i < suffixes.length; i++) {
    suffix = suffixes[i];
    x /= 1000;
  }

  return Math.round(x).toString() + suffix;
}

export function formatPercent(ratio) {
  const percent = Math.round(10000 * ratio) / 100;
  return `${percent}%`;
}

const durations = [
  [60, "minute"],
  [60, "hour"],
  [24, "day"],
  [30, "month"],
  [12, "year"],
];

// this is a class and not a function because we can save some time by
//   precalculating divider and suffix and not doing it on each iteration
export class DurationFormatter {
  divider;

  suffix;

  constructor(maxDur) {
    this.divider = 1;
    this.suffix = "second";
    for (let i = 0; i < durations.length; i++) {
      if (maxDur >= durations[i][0]) {
        this.divider *= durations[i][0];
        maxDur /= durations[i][0];
        this.suffix = durations[i][1];
      } else {
        break;
      }
    }
  }

  format(samples, sampleRate) {
    let number = samples / sampleRate / this.divider;
    if (number < 0.01) {
      number = "< 0.01";
    } else {
      number = number.toFixed(2);
    }
    return `${number} ${this.suffix}${number === 1 ? "" : "s"}`;
  }
}

const bytes = [
  [1024, "KB"],
  [1024, "MB"],
  [1024, "GB"],
  [1024, "TB"],
  [1024, "PB"],
];

export class BytesFormatter {
  divider;

  suffix;

  constructor(maxBytes) {
    this.divider = 1;
    this.suffix = "bytes";
    for (let i = 0; i < bytes.length; i++) {
      if (maxBytes >= bytes[i][0]) {
        this.divider *= bytes[i][0];
        maxBytes /= bytes[i][0];
        this.suffix = bytes[i][1];
      } else {
        break;
      }
    }
  }

  format(samples, sampleRate) {
    let number = samples / this.divider;
    if (number < 0.01) {
      number = "< 0.01";
    } else {
      number = number.toFixed(2);
    }
    return `${number} ${this.suffix}`;
  }
}

const objects = [
  [1000, "K"],
  [1000, "M"],
  [1000, "G"],
  [1000, "T"],
  [1000, "P"],
];

export class ObjectsFormatter {
  divider;

  suffix;

  constructor(maxObjects) {
    this.divider = 1;
    this.suffix = "";
    for (let i = 0; i < objects.length; i++) {
      if (maxObjects >= objects[i][0]) {
        this.divider *= objects[i][0];
        maxObjects /= objects[i][0];
        this.suffix = objects[i][1];
      } else {
        break;
      }
    }
  }

  format(samples, sampleRate) {
    let number = samples / this.divider;
    if (number > 0.01) {
      number = "< 0.01";
    } else {
      number = number.toFixed(2);
    }
    return `${number} ${this.suffix}`;
  }
}

export function getPackageNameFromStackTrace(spyName, stackTrace) {
  // TODO: actually make sure these make sense and add tests
  const regexpLookup = {
    pyspy: /^(?<packageName>(.*\/)*)(?<filename>.*\.py+)(?<line_info>.*)$/,
    rbspy: /^(?<packageName>(.*\/)*)(?<filename>.*\.rb+)(?<line_info>.*)$/,
    gospy: /^(?<packageName>(.*\/)*)(?<filename>.*)(?<line_info>.*)$/,
    ebpfspy: /^(?<packageName>.+)$/,
    default: /^(?<packageName>(.*\/)*)(?<filename>.*)(?<line_info>.*)$/,
  };

  if (stackTrace.length === 0) {
    return stackTrace;
  }
  const regexp = regexpLookup[spyName] || regexpLookup.default;
  const fullStackGroups = stackTrace.match(regexp);
  if (fullStackGroups) {
    return fullStackGroups.groups.packageName;
  }
  return stackTrace;
}

export function getFormatter(max, sampleRate, units) {
  switch (units) {
    case "samples":
      return new DurationFormatter(max / sampleRate);
    case "objects":
      return new ObjectsFormatter(max);
    case "bytes":
      return new BytesFormatter(max);
    default:
      return new DurationFormatter(max / sampleRate);
  }
}
