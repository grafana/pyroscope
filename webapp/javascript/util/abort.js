export function isAbortError(err) {
  if (!err) {
    return false;
  }

  // https://developer.mozilla.org/en-US/docs/Web/API/DOMException
  return err.name === 'AbortError' || err.code === 20;
}
