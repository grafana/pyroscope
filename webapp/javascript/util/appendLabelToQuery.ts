export function appendLabelToQuery(
  query: string,
  label: string,
  labelValue: string
) {
  const case1Regexp = new RegExp(`${label}=.+?(\\}|,)`);
  if (query.match(case1Regexp)) {
    return query.replace(case1Regexp, `${label}="${labelValue}"$1`);
  }
  if (query.indexOf('{}') !== -1) {
    return query.replace('}', `${label}="${labelValue}"}`);
  }
  if (query.indexOf('}') !== -1) {
    return query.replace('}', `, ${label}="${labelValue}"}`);
  }

  console.warn('TODO: handle this case');
  return query;
}
