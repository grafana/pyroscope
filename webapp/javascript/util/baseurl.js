// There's a copy of this function in welcome.html
//   TODO: maybe dedup somehow?
function basename() {
  const baseURLMetaTag = document.querySelector(
    'meta[name="pyroscope-base-url"]'
  );
  if (!baseURLMetaTag) {
    return null;
  }

  const baseURL = baseURLMetaTag.content;

  if (!baseURL) {
    return null;
  }
  const url = new URL(baseURL, window.location.href);
  return url.pathname;
}

export default basename;
