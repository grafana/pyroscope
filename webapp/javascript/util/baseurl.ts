// There's a copy of this function in welcome.html
//   TODO: maybe dedup somehow?
function basename() {
  const baseURLMetaTag = document.querySelector(
    'meta[name="pyroscope-base-url"]'
  ) as HTMLMetaElement;

  if (!baseURLMetaTag) {
    return undefined;
  }

  const baseURL = baseURLMetaTag.content;

  if (!baseURL) {
    return undefined;
  }
  const url = new URL(baseURL, window.location.href);
  return url.pathname;
}

export default basename;
