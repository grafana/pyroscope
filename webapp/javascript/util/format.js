export function numberWithCommas(x) {
  return x.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');
}

const suffixes = [
  "K",
  "M",
  "G",
  "T",
];

export function shortNumber(x) {
  let suffix = '';

  for(var i = 0; x > 1000 && i < suffixes.length; i++) {
    suffix = suffixes[i];
    x /= 1000;
  }

  return Math.round(x).toString() + suffix;
}

export function formatPercent(ratio) {
  const percent = Math.round(10000 * ratio) / 100;
  return percent+'%';
}

const durations = [
  [60, "minute"],
  [60, "hour"],
  [24, "day"],
  [30, "month"],
  [12, "year"],
];

// this is a class and not a function because we can save some time by
//   precalculating divider and suffix and not doing it on each iteration
export class DurationFormater {
  constructor(maxDur) {
    this.divider = 1;
    this.suffix = 'second';
    for(var i = 0; i < durations.length; i++) {
      if (maxDur >= durations[i][0]) {
        this.divider *= durations[i][0];
        maxDur /= durations[i][0];
        this.suffix = durations[i][1];
      } else {
        break;
      }
    }
  }

  format(seconds) {
    let number = seconds / this.divider;
    if (number < 0.01) {
      number = '< 0.01';
    } else {
      number = number.toFixed(2);
    }
    return `${number} ${this.suffix}` + (number == 1 ? '' : 's');
  }
}


export function getPackageNameFromStackTrace(spyName, stackTrace) {
  // TODO: actually make sure these make sense and add tests
  const regexpLookup = {
    pyspy: /^(?<packageName>(.*\/)*)(?<filename>.*\.py+)(?<line_info>.*)$/,
    rbspy: /^(?<packageName>(.*\/)*)(?<filename>.*\.rb+)(?<line_info>.*)$/,
    gospy: /^(?<packageName>(.*\/)*)(?<filename>.*)(?<line_info>.*)$/,
    default: /^(?<packageName>(.*\/)*)(?<filename>.*)(?<line_info>.*)$/,
  };

  if (stackTrace.length === 0) {
    return stackTrace;
  }
  const regexp = regexpLookup[spyName] || regexpLookup.default;
  const fullStackGroups = stackTrace.match(regexp);
  if (fullStackGroups) {
    return fullStackGroups.groups.packageName;
  }
  return stackTrace;
}

