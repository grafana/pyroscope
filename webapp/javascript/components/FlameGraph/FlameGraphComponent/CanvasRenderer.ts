import { PX_PER_LEVEL, COLLAPSE_THRESHOLD, BAR_HEIGHT, GAP } from './constants';
import { formatPercent, getFormatter, Units } from '../../../util/format';
import { fitToCanvasRect } from '../../../util/fitMode';

export interface CanvasRendererConfig {
  //  context: CanvasRenderingContext2D;
  canvas: HTMLCanvasElement;
  numTicks: number;
  sampleRate: number;
  names: string[];
  levels: number[][];

  viewType: 'single' | 'diff';

  pxPerTick: number;

  topLevel: number;
  rangeMin: number;
  rangeMax: number;

  units: Units;
  fitMode: 'TAIL' | 'HEAD'; // TODO import from fitMode
}

// TODO
// this shouldn't really be a component
// so don't call it props
export function RenderCanvas(props: CanvasRendererConfig) {
  const { canvas } = props;
  const { numTicks, rangeMin, rangeMax, sampleRate } = props;
  const { fitMode } = props;
  const { units } = props;

  const graphWidth = canvas.clientWidth;
  // TODO: why is this needed? otherwise height is all messed up
  canvas.width = graphWidth;

  const pxPerTick = graphWidth / numTicks / (rangeMax - rangeMin);

  const ctx = canvas.getContext('2d');

  // TODO what does ff mean?
  const { viewType } = props;
  const ff = createFF(viewType);

  const { names, levels, topLevel } = props;
  const formatter = getFormatter(numTicks, sampleRate, units);

  for (let i = 0; i < levels.length - topLevel; i += 1) {
    const level = levels[topLevel + i];
    for (let j = 0; j < level.length; j += ff.jStep) {
      const barIndex = ff.getBarOffset(level, j);
      const numBarTicks = ff.getBarTotal(level, j);

      const x = tickToX(numTicks, rangeMin, pxPerTick, barIndex);
      const y = i * PX_PER_LEVEL;

      // merge very small blocks into big "collapsed" ones for performance
      const collapsed = numBarTicks * pxPerTick <= COLLAPSE_THRESHOLD;

      const sw = numBarTicks * pxPerTick - (collapsed ? 0 : GAP);
      const sh = BAR_HEIGHT;

      // Decide the name
      const ratio = numBarTicks / numTicks;

      const percent = formatPercent(ratio);
      const shortName = names[level[j + ff.jName]];
      const longName = `${shortName} (${percent}, ${formatter.format(
        numBarTicks,
        sampleRate
      )})`;

      const namePosX = Math.round(Math.max(x, 0));

      // It's important to set the font before hand
      // Since it will be used to calculate how many characters can fit
      ctx.textBaseline = 'middle';
      ctx.font =
        '400 11.5px SFMono-Regular, Consolas, Liberation Mono, Menlo, monospace';
      // Since this is a monospaced font any character would do
      const characterSize = ctx.measureText('a').width;

      const fitCalc = fitToCanvasRect({
        mode: fitMode,
        charSize: characterSize,
        rectWidth: sw,
        fullText: longName,
        shortText: shortName,
      });

      // Draw the block
      ctx.beginPath();
      ctx.rect(x, y, sw, sh);
      // TODO color
      ctx.fillStyle = '#48CE73'; // green
      ctx.fill();

      ctx.save();
      ctx.clip();
      ctx.fillStyle = 'black';
      // when showing the code, give it a space in the beginning
      ctx.fillText(fitCalc.text, namePosX + fitCalc.marginLeft, y + sh / 2 + 1);
      ctx.restore();
    }
  }
}

function createFF(viewType: CanvasRendererConfig['viewType']) {
  switch (viewType) {
    case 'single': {
      return formatSingle;
    }
    case 'diff': {
      return formatDouble;
    }

    default:
      throw new Error(`Format not supported: '${viewType}'`);
  }
}

const formatSingle = {
  format: 'single',
  jStep: 4,
  jName: 3,
  getBarOffset: (level: number[], j: number) => level[j],
  getBarTotal: (level: number[], j: number) => level[j + 1],
  getBarTotalDiff: (level: number[], j: number) => 0,
  getBarSelf: (level: number[], j: number) => level[j + 2],
  getBarSelfDiff: (level: number[], j: number) => 0,
  getBarName: (level: number[], j: number) => level[j + 3],
};

const formatDouble = {
  format: 'double',
  jStep: 7,
  jName: 6,
  getBarOffset: (level: number[], j: number) => level[j] + level[j + 3],
  getBarTotal: (level: number[], j: number) => level[j + 4] + level[j + 1],
  getBarTotalLeft: (level: number[], j: number) => level[j + 1],
  getBarTotalRght: (level: number[], j: number) => level[j + 4],
  getBarTotalDiff: (level: number[], j: number) => {
    return level[j + 4] - level[j + 1];
  },
  getBarSelf: (level: number[], j: number) => level[j + 5] + level[j + 2],
  getBarSelfLeft: (level: number[], j: number) => level[j + 2],
  getBarSelfRght: (level: number[], j: number) => level[j + 5],
  getBarSelfDiff: (level: number[], j: number) => level[j + 5] - level[j + 2],
  getBarName: (level: number[], j: number) => level[j + 6],
};

function tickToX(
  numTicks: number,
  rangeMin: number,
  pxPerTick: number,
  i: number
) {
  return (i - numTicks * rangeMin) * pxPerTick;
}
