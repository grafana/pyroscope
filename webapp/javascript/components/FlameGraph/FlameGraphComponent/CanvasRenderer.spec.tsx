/* eslint-disable no-underscore-dangle */
import { Units } from '../../../util/format';
import { RenderCanvas } from './CanvasRenderer';

describe('CanvasRenderer', () => {
  let canvas: HTMLCanvasElement;
  let ctx: CanvasRenderingContext2D;
  beforeEach(() => {
    canvas = document.createElement('canvas');
    canvas.width = 400;
    canvas.height = 400;
    ctx = canvas.getContext('2d');
  });

  it('works', () => {
    RenderCanvas({
      canvas,
      topLevel: 0,
      rangeMin: 0,
      pxPerTick: 0.43724696356275305,
      viewType: 'single',
      numTicks: 988,
      sampleRate: 100,
      names: [
        'total',
        'runtime.main',
        'main.slowFunction',
        'main.work',
        'main.main',
        'main.fastFunction',
      ],
      levels: [
        [0, 988, 0, 0],
        [0, 988, 0, 1],
        [0, 214, 0, 5, 214, 3, 2, 4, 217, 771, 0, 2],
        [0, 214, 214, 3, 216, 1, 1, 5, 217, 771, 771, 3],
      ],

      rangeMax: 0,
      units: Units.Samples,
      fitMode: 'HEAD',
    });

    expect(ctx.__getEvents()).toMatchSnapshot();
  });
});
