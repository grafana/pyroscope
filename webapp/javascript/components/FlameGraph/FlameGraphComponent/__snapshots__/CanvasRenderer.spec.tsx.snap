// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CanvasRenderer works 1`] = `
Array [
  Object {
    "props": Object {
      "value": "middle",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textBaseline",
  },
  Object {
    "props": Object {
      "value": "11.5px SFMono-Regular, Consolas, \\"Liberation Mono\\", Menlo, monospace",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  Object {
    "props": Object {
      "text": "a",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "value": "#48ce73",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "save",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clip",
  },
  Object {
    "props": Object {
      "value": "#000000",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "restore",
  },
  Object {
    "props": Object {
      "value": "middle",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textBaseline",
  },
  Object {
    "props": Object {
      "value": "11.5px SFMono-Regular, Consolas, \\"Liberation Mono\\", Menlo, monospace",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  Object {
    "props": Object {
      "text": "a",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "value": "#48ce73",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "save",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clip",
  },
  Object {
    "props": Object {
      "value": "#000000",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "restore",
  },
  Object {
    "props": Object {
      "value": "middle",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textBaseline",
  },
  Object {
    "props": Object {
      "value": "11.5px SFMono-Regular, Consolas, \\"Liberation Mono\\", Menlo, monospace",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  Object {
    "props": Object {
      "text": "a",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "value": "#48ce73",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "save",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clip",
  },
  Object {
    "props": Object {
      "value": "#000000",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "restore",
  },
  Object {
    "props": Object {
      "value": "middle",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textBaseline",
  },
  Object {
    "props": Object {
      "value": "11.5px SFMono-Regular, Consolas, \\"Liberation Mono\\", Menlo, monospace",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  Object {
    "props": Object {
      "text": "a",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "value": "#48ce73",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "save",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clip",
  },
  Object {
    "props": Object {
      "value": "#000000",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "restore",
  },
  Object {
    "props": Object {
      "value": "middle",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textBaseline",
  },
  Object {
    "props": Object {
      "value": "11.5px SFMono-Regular, Consolas, \\"Liberation Mono\\", Menlo, monospace",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  Object {
    "props": Object {
      "text": "a",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "value": "#48ce73",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "save",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clip",
  },
  Object {
    "props": Object {
      "value": "#000000",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "restore",
  },
  Object {
    "props": Object {
      "value": "middle",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textBaseline",
  },
  Object {
    "props": Object {
      "value": "11.5px SFMono-Regular, Consolas, \\"Liberation Mono\\", Menlo, monospace",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  Object {
    "props": Object {
      "text": "a",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "value": "#48ce73",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "save",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clip",
  },
  Object {
    "props": Object {
      "value": "#000000",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "restore",
  },
  Object {
    "props": Object {
      "value": "middle",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textBaseline",
  },
  Object {
    "props": Object {
      "value": "11.5px SFMono-Regular, Consolas, \\"Liberation Mono\\", Menlo, monospace",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  Object {
    "props": Object {
      "text": "a",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "value": "#48ce73",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "save",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clip",
  },
  Object {
    "props": Object {
      "value": "#000000",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "restore",
  },
  Object {
    "props": Object {
      "value": "middle",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textBaseline",
  },
  Object {
    "props": Object {
      "value": "11.5px SFMono-Regular, Consolas, \\"Liberation Mono\\", Menlo, monospace",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  Object {
    "props": Object {
      "text": "a",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "value": "#48ce73",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "save",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clip",
  },
  Object {
    "props": Object {
      "value": "#000000",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "restore",
  },
]
`;
